### DISCUSSION TITLE
segmentation fault using libcurl in multithreaded process

### DISCUSSION THREAD

**Discussion by jorgeluislazo** [curl/curl#14716] [Q&A] [answered]
Hi,

I have a rather complex Perl program using libcurl directly that turns into sporadic segmentation faults after several hours of workload. The stack trace always looks like this:

```
                Stack trace of thread 1201532:
                #0  0x00007fe99e4b2df8 Perl_csighandler (libperl.so.5.26)
                #1  0x00007fe99e1dbd80 __restore_rt (libpthread.so.0)
                #2  0x00007fe99d35ab41 __poll (libc.so.6)
                #3  0x00007fe99dfb9e79 __res_context_send (libresolv.so.2)
                #4  0x00007fe99dfb77cf __res_context_query (libresolv.so.2)
                #5  0x00007fe99dfb7e76 __res_context_querydomain (libresolv.so.2)
                #6  0x00007fe99dfb846d __res_context_search (libresolv.so.2)
                #7  0x00007fe99459515e _nss_dns_gethostbyname4_r (libnss_dns.so.2)
                #8  0x00007fe99d34eb9e gaih_inet.constprop.6 (libc.so.6)
                #9  0x00007fe99d34fe5b getaddrinfo (libc.so.6)
                #10 0x00007fe9997a7cc4 Curl_getaddrinfo_ex (libcurl.so.4)
                #11 0x00007fe9997b17e1 getaddrinfo_thread (libcurl.so.4)
                #12 0x00007fe9997af00b curl_thread_create_thunk (libcurl.so.4)
                #13 0x00007fe99e1d11da start_thread (libpthread.so.0)
                #14 0x00007fe99d2618d3 __clone (libc.so.6)
                
                Stack trace of thread 1183917:
                #0  0x00007fe99e4e1397 Perl_sv_upgrade (libperl.so.5.26)
                #1  0x00007fe99e4dfd4b Perl_sv_setsv_flags (libperl.so.5.26)
                #2  0x00007fe99e4c90c8 Perl_av_make (libperl.so.5.26)
                #3  0x00007fe99e502a91 Perl_pp_anonlist (libperl.so.5.26)
                #4  0x00007fe99e4ca315 Perl_runops_standard (libperl.so.5.26)
                #5  0x00007fe99e50a941 S_docatch (libperl.so.5.26)
                #6  0x00007fe99e4ca315 Perl_runops_standard (libperl.so.5.26)
                #7  0x00007fe99e441f2d Perl_call_sv (libperl.so.5.26)
                #8  0x00007fe999e1b14b poe_data_ev_dispatch_due (EPoll.so)
                #9  0x00007fe999e1a420 lp_loop_run (EPoll.so)
                #10 0x00007fe99e4d24a9 Perl_pp_entersub (libperl.so.5.26)
                #11 0x00007fe99e4ca315 Perl_runops_standard (libperl.so.5.26)
                #12 0x00007fe99e44a10f perl_run (libperl.so.5.26)
                #13 0x000055aa13e00eda main (perl)
                #14 0x00007fe99d2628a5 __libc_start_main (libc.so.6)
                #15 0x000055aa13e00f1e _start (perl)
```
Now, there may be other reasons as to why this segfault happens, but I want to ask for help if there is anything obvious in the stacktrace related to libcurl that I can try to debug on.

here is my version of curl and libcurl:

```
curl 7.64.1 (x86_64-pc-linux-gnu) libcurl/7.64.1 OpenSSL/1.1.1k zlib/1.2.11
Release-Date: 2019-03-27
Protocols: dict file ftp ftps gopher http https imap imaps ldap ldaps pop3 pop3s rtsp smb smbs smtp smtps telnet tftp 
Features: AsynchDNS HTTPS-proxy IPv6 Largefile libz NTLM NTLM_WB SSL TLS-SRP UnixSockets

```

Any ideas are appreciated

**Comment by dfandrich** [comment, marked as answer]
It looks like your libcurl is using the --enable-threaded-resolver. This
creates a thread for name resolves which introduces all the issues with signals
and POSIX threads. Are your program's signal handlers prepared for that?  A
brute-force hammer to try would be to switch to the c-ares name resolver which
does not create any threads. Also, note that your libcurl version is over 5
years out of date.

**Reply by jorgeluislazo** [reply]
I thought the program signal handling may not be an issue since this exact code had been running in an older Linux environment for a long time without issues. This problem started happening after we had to re-install the perl program in a newer Linux environment using Oracle Linux 8 (although the default OL8 libcurl is ancient). I know this environment switch could have introduced a lot of different factors, but at the very least its the same libcurl version/features.

Should I try the same version rebuilt with c-ares? or a much newer libcurl and also with c-ares? I am not familiar how much the threaded resolver has changed with different versions of libcurl

**Reply by bagder** [reply]
(the signal seems to come from resolver code, not from libcurl)

I think the most helpful things to do in order of importance:

1. build and use a modern libcurl
2. enable debug in the build so that the debug symbols are kept as they offer more insights into stack traces
3. work on a recipe for others to reproduce the problem so that we can aid with debugging

**Reply by jorgeluislazo** [reply]
Thank you for the insight, I rebuilt and installed libcurl 7.88 with c-ares/1.33.1 this time, and that did the trick. I think Dan is right in that threaded resolver was the issue, although I'm not sure why the older perl code library (or environment) was able to handle it just fine on the same work load.

Either way, I hope that anybody that finds a similar stacktrace can find this discussion helpful if your code is interfacing with curl_multi_* functions the same way my code is.
